KAFKA-10023: Enforce broker-wide and per-listener connection creation…

apovzner opened the pull request:
This PR implements the part of KIP-612 that adds broker configurations for broker-wide and per-listener connection creation rate limits and enforces these limits.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apovzner committed:
KAFKA-10023: Enforce broker-wide and per-listener connection creation rate

dajac commented on clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java:
```diff
@@ -161,6 +162,9 @@ private KafkaMbean removeAttribute(KafkaMetric metric, String mBeanName) {
     private String addAttribute(KafkaMetric metric) {
         try {
             MetricName metricName = metric.metricName();
+            if (metricName.tags().containsKey(DO_NOT_REPORT_TAG)) {
+                return null;
+            }
```
I am not convinced by this. The main issue being that other reporters will report the metric. If we really want to not report a metric, I think that we need a solution which works for all reporters. Could you perhaps elaborate more on the need here?

I can think of the following alternatives:
* add a flag to the sensor to indicate if it must be reported or not.
* don't rely on metrics to create/store the sensor but have a local reference.

apovzner replied:
You are right, I replaced this with your first alternative.

dajac commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -689,6 +691,11 @@ object KafkaConfig {
+    s"may also be configured by prefixing the config name with the listener prefix, for example, <code>listener.name.internal.$MaxConnectionCreationRateProp</code>." +
+    "Broker-wide connection rate limit should be configured based on broker capacity while listener limits should be configured based on " +
+    "application requirements. New connections will be throttled if either the listener or the broker limit is reached, with the exception " +
+    " of inter-broker listener. Connections on the inter-broker listener will be throttled only when the listener-level rate limit is reached. "
```
nit: extra space after the `.`.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1256,11 +1272,17 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
         do {
-          counts.wait()
-        } while (!connectionSlotAvailable(listenerName))
+          counts.wait(remainingThrottleTimeMs)
```
A thread waiting here will be notified when a connection is closed (when `dec` is called). As connections in AK are long lived, couldn't we end up in a case where a connection is throttled for a longer period than the computed `trottleTimeMs` if no connection is closed in between?

apovzner replied:
This is exactly the behavior proposed in KIP -- if we reach any limit (number of connections or connection rate), we need to wait. So, if there is no space for a new connection, and the delay due to rate limit has passed, we would have to wait for a connection slot. However, remember that if we are waiting on an inter-broker connection slot, the broker finds and closes a connection of another listener to accommodate inter-broker connection. See KIP-402.

dajac replied:
Thanks for the clarification. I am sorry but I misread the code the first time.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1287,15 +1309,97 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+      listenerThrottleTimeMs
+    } else {
+      val brokerThrottleTimeMs = recordAndGetThrottleTimeMs(brokerConnectionRateSensor, timeMs)
+      val throttleTimeMs = math.max(brokerThrottleTimeMs, listenerThrottleTimeMs)
```
nit: This variable is not really needed.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1306,18 +1410,26 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
     override def reconfigure(configs: util.Map[String, _]): Unit = {
       lock.synchronized {
         _maxConnections = maxConnections(configs)
+        updateConnectionRateQuota(maxConnectionCreationRate(configs), Some(listener.value()))
```
nit: `()` can be removed after `value`.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1306,18 +1410,26 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
     private def maxConnections(configs: util.Map[String, _]): Int = {
       Option(configs.get(KafkaConfig.MaxConnectionsProp)).map(_.toString.toInt).getOrElse(Int.MaxValue)
     }
+    private def maxConnectionCreationRate(configs: util.Map[String, _]): Int = {
```
nit: I would put an empty line before declaring the method.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1287,15 +1309,97 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+
+  private def connectionRateMetricName(listenerOpt: Option[String]): MetricName = {
+    val quotaEntity = listenerOpt.getOrElse("broker")
+    metrics.metricName(
+      s"connection-creation-rate-$quotaEntity",
+      "connection-quota-no-jmx",
+      s"Tracking $quotaEntity connection creation rate",
+      rateQuotaMetricTags(listenerOpt))
```
Wouldn't it make sense to expose the metrics? I think that they could be useful to know if connections are throttled at the broker or at the listener level, no? Moreover, the number of metrics is small so it should not hurt.

apovzner replied:
We already have a yammer Meter that reports connection creation rate, also tagged with listener and processor. This same meter tracks both connection creation rate and total. Here is connetion creation rate metric:
kafka.server:type=socket-server-metrics,listener={listener_name},networkProcessor={#},name=connection-creation-rate

I had to add a Rate metric here because it both tracks rate and quota, and allows to calculate throttle time the same way we do with client (request & bandwidth) quotas.

dajac replied:
Ack. I did not know this.

dajac commented on core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala:
```diff
@@ -163,13 +167,65 @@ class DynamicConnectionQuotaTest extends BaseRequestTest {
+
+    // Set 7 conn/sec rate limit for each listener and verify it gets enforced
+    val plaintextListenerProp = s"${listener.configPrefix}${KafkaConfig.MaxConnectionCreationRateProp}"
+    props.put(s"listener.name.external.${KafkaConfig.MaxConnectionCreationRateProp}", "7")
```
nit: Could we define a val for `7` and reuse it bellow as well?

dajac reviewed (commented):
@apovzner Thanks for the PR! I just made a first pass over it and left few comments/questions.

apovzner committed:
fixed bad merge

apovzner committed:
addressed initial review

dajac commented on clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java:
```diff
@@ -572,16 +572,18 @@ public synchronized void removeReporter(MetricsReporter reporter) {
     }
 
-    synchronized void registerMetric(KafkaMetric metric) {
+    synchronized void registerMetric(KafkaMetric metric, boolean report) {
```
Thinking a bit more about this, did you consider adding the flag to `MetricConfig`? It may be a bit simpler and cleaner as it avoids having to add the flag to all the methods. What do you think?

apovzner replied:
That's a good idea I haven't thought about. I tried that and it does look better. Will update PR with this approach.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1258,11 +1274,17 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
                                     acceptorBlockedPercentMeter: com.yammer.metrics.core.Meter): Unit = {
     counts.synchronized {
-      if (!connectionSlotAvailable(listenerName)) {
+      val startTimeMs = time.milliseconds()
```
nit: `()` can be omitted here.

dajac replied:
It is a bit confusing to have `startTimeMs` and `startNs` defined few lines apart. Is it worth renaming `startTimeMs` to `startThrottleTimeMs` to clearly state that this is used as part of the throttle time computing. It would also be consistent with `endThrottleTimeMs`.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,95 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+  }
+
+  private def rateQuotaMetricTags(listenerOpt: Option[String]): util.Map[String, String] = {
+    val tags = new util.LinkedHashMap[String, String]
+    listenerOpt.foreach(listener => tags.put("listener", listener))
+    tags
```
What about returning `Collections.emptyMap` when `listenerOpt` is not defined and using `Collections.singletonMap` when it is?

apovzner replied:
I realized that I don't need tags here anymore, since the name of the metric contains the name of the listener, so the metrics are already distinct per listener (and broker-wide). I removed that method altogether.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,95 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+    val quotaEntity = listenerOpt.getOrElse("broker")
+    metrics.metricName(
+      s"connection-creation-rate-$quotaEntity",
+      "connection-quota-no-jmx",
```
nit: I would rename this one now.

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,95 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+  }
+
+  private def updateConnectionRateQuota(quotaLimit: Int, listenerOpt: Option[String] = None): Unit = {
+    val metric = metrics.metric(connectionRateMetricName((listenerOpt)))
```
nit: Are the parenthesis around `listenerOpt` really necessary?

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,95 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+  }
+
+  private def connectionRateMetricName(listenerOpt: Option[String]): MetricName = {
+    val quotaEntity = listenerOpt.getOrElse("broker")
```
nit: As the `quotaEntity` is also computed by the caller methods, would it make sense to pass it as an argument to `connectionRateMetricName`?

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1308,18 +1410,27 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
       val value = maxConnections(configs)
       if (value <= 0)
         throw new ConfigException("Invalid max.connections $listenerMax")
```
Not related to your PR but it seems that `listenerMax` is never defined. I think that `value` should be used instead here. Could you fix this? Could you also use `MaxConnectionsProp` instead of `max.connections` as you did for the other already?

apovzner replied:
yes, good catch. Fixed.

dajac commented on core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala:
```diff
@@ -163,13 +167,66 @@ class DynamicConnectionQuotaTest extends BaseRequestTest {
+    // before setting connection rate to 10, verify we can do at least double that by default (no limit)
+    verifyConnectionRate(2 * connRateLimit, Int.MaxValue, "PLAINTEXT")
+
+    // Reduce total broker connection rate limit to 10 at the cluster level and verify the limit is enforced
```
`to 18`?

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -329,15 +613,46 @@ class ConnectionQuotasTest {
+    acceptConnections(connectionQuotas, listenerDesc.listenerName, listenerDesc.defaultIp, numConnections, timeIntervalMs)
+    val elapsedSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis - startTimeMs)
+    val actualRate = (numConnections.toDouble / elapsedSeconds).toInt
+    if (actualRate - epsilon > expectedRate || actualRate + epsilon < expectedRate)
+      throw new TestFailedException(
+        (e: StackDepthException) =>
+          Some(s"Expected rate $expectedRate, but got $actualRate ($numConnections connections / $elapsedSeconds sec)"),
+        None,
+        Position.here)
```
Can't we use `assertTrue` here to simplify?

apovzner replied:
Yes, I had a memory that asserting from another thread would not do the right thing, but I think that was a very old memory. Checked and it works well. Replaced with `assertTrue`.

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+  }
+
+  @Test
+  def testMaxListenerConnectionRateReconfiguration(): Unit = {
```
I wonder if we could just check that the metric's config is correctly re-configured instead of testing the number of connections accepted. The goal of the test is not really to verify that the quota works but rather to ensure that metric is correctly re-configured. Have you considered this? The same would apply to `testMaxBrokerConnectionRateReconfiguration`.

apovzner replied:
I still think it's important to test all the way that it does throttle to the correct value. We had a bug pretty recently, KAFKA-9658, where adding and then removing bandwidth quotas via a dynamic config changed the config and even changed some data structs, but not all data structs to actually change the throttling behavior. So, I feel more confident keeping the test up to checking the throttle.

dajac replied:
Fair enough.

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // create connections with the total rate < broker-wide quota, and verify there is no throttling
+      val connCreateIntervalMs = 25 // connection creation rate = 40/sec per listener (3 * 40 = 120/sec total)
+      val connectionsPerListener = 200 // should take 5 seconds to create 200 connections with rate = 40/sec
```
I really like the explanation next to the constants! I would recommend to group all the constants in the beginning of the test case. That would help to get a quick overview of the test case.

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+  @Test
+  def testBrokerConnectionRateLimitWhenActualRateAboveLimit(): Unit = {
+    val brokerRateLimit = 90
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionCreationRateProp, brokerRateLimit.toString)
+    val config = KafkaConfig.fromProps(props)
```
nit: We could perhaps create a small help method like `brokerProps` that accepts a Map of customer configuration pairs and returns a `KafkaConfig`. That would reduce the boilerplate code.

apovzner replied:
the test code has just one extra prop in addition to `brokerPropsWithDefaultConnectionLimits` -- somehow it is easier to read the current way. I will keep it.

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
```
It seems that almost all the test cases instantiate an executor with `listeners.size`. Have you considered moving this to the `setUp` method and moving the `shutdownNow` to the `tearDown`?

apovzner replied:
others use single thread, but yes, makes sense just to create one upfront using `listeners.size`. Will do that.

dajac commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -302,21 +312,295 @@ class ConnectionQuotasTest {
+      futures.foreach(_.get(30, TimeUnit.SECONDS))
+
+      // verify that every listener was throttled
+      blockedPercentMeters.foreach { case (name, meter) =>
+        assertTrue(s"Expected BlockedPercentMeter metric for $name listener to be recorded", meter.count() > 0)
+      }
```
nit: Would it make sense to add the pendant of `verifyNoBlockedPercentRecordedOnAllListeners` for this block? Something like `verifyNonZeroBlockedPercentRecordedOnAllListeners`?

dajac reviewed (commented):
@apovzner Thanks for the updated PR. The overall logic looks good. I have left few more comments.

apovzner committed:
moved no reporting flag to MetricConfig and addressed other comments

apovzner committed:
reverted unintended changes (removed spaces in a comment)

apovzner committed:
one more revert

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,89 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+    }
+  }
+
+  private def recordAndGetThrottleTimeMs(sensor: Sensor, timeMs: Long): Int = {
```
nit: Shall we add a javadoc to this one as well?

dajac commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,89 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+    sensor
+  }
+
+  private def updateConnectionRateQuota(quotaLimit: Int, listenerOpt: Option[String] = None): Unit = {
```
nit: Shall we add a javadoc to this one as well?

dajac reviewed (commented):
@apovzner Thanks for the update. I just left two cosmetic comments. LGTM pending Jenkins.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/metrics/MetricConfig.java:
```diff
@@ -97,5 +99,13 @@ public MetricConfig recordLevel(Sensor.RecordingLevel recordingLevel) {
         return this;
     }
 
+    public boolean skipReporting() {
```
We don't talk about this in the KIP and this is a public class. Why are we skipping reporting anyway?

dajac replied:
I asked the very same question: https://github.com/apache/kafka/pull/8768#discussion_r433705077. Basically, we already have the same metrics defined as part of the `SelectorMetrics` so the ones added here collides.

Thinking a bit more about this, I wonder if we could use Sensors without adding them to the registry at all.

rajinisivaram replied:
If we are using a sensor to determine throttle time that is different from the one in Selector, we might want to expose  it as a metric anyway. In case of a bug, we want to know this metric, not one in Selector. Perhaps we could use `connections-accepted` instead of `connections-created` or something like that. In any case, `skipReporting` seems odd, so as @dajac  said, using a Sensor that is not added to the metrics registry may be an option too.

rajinisivaram commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,89 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+  }
+
+  private def updateConnectionRateQuota(quotaLimit: Int, listenerOpt: Option[String] = None): Unit = {
+    val quotaEntity = listenerOpt.getOrElse("broker")
```
`broker` is a valid listener name, perhaps a reasonable name for inter-broker listener.

rajinisivaram commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1289,15 +1311,89 @@ class ConnectionQuotas(config: KafkaConfig, time: Time) extends Logging {
+    private val _connectionRateSensor = createConnectionRateQuotaSensor(Int.MaxValue, Some(listener.value))
 
     def maxConnections: Int = _maxConnections
+    def connectionRateSensor: Sensor = _connectionRateSensor
```
We could just make `connectionRateSensor` a `val`

rajinisivaram commented on core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala:
```diff
@@ -70,37 +80,38 @@ class ConnectionQuotasTest {
     }
+    // use system time, because ConnectionQuota causes the current thread to wait with timeout, which waits based on
+    // system time; so using mock time will likely result in test flakiness due to a mixed use of mock and system time
+    metrics = new Metrics(new MetricConfig(), Collections.emptyList(), Time.SYSTEM)
```
Is this required for all tests or only the new connection rate tests?

apovzner replied:
This is required for all tests. For tests that are not supposed to trigger throttling due to connection rate quota, we want this because if the code incorrectly throttles to limit rate (calls wait() with timeout), the existing tests may start failing in a way that is hard to debug (timeout too early or too late, not in a place we expect).

rajinisivaram reviewed (commented):
@apovzner Thanks for the PR, looks good, left a few comments/questions.

rajinisivaram commented:
ok to test

rajinisivaram commented:
There is one test failure in `kafka.network.DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota` that is related to the PR.

apovzner committed:
Fixed merge issue, reporting connection-accept-rate metrics

apovzner commented:
@rajinisivaram Thanks for your review.  I updated this PR to expose  'connection-accept-rate'  metrics, addressed the remaining comments, and the test should also be fixed.

It seems like it would also make sense to add and expose throttle time metrics. What do you think? If so, I will have a separate PR to add throttle-time metrics and use Token Bucket for quota calculation that David added as part of KIP-599 that works better with bursty workloads. I will update the KIP accordingly and notify on voting thread.

rajinisivaram reviewed (approved):
@apovzner Thanks for the updates, LGTM. Yes, adding throttle time metrics and using token bucket for quota calculation in a follow-on PR sounds good.

rajinisivaram commented:
ok to test

rajinisivaram commented:
@apovzner DynamicBrokerReconfigurationTest.testAddRemoveSaslListeners failed in all three PR builds, so probably related?
```
15:55:45 kafka.server.DynamicBrokerReconfigurationTest > testAddRemoveSaslListeners FAILED
15:55:45     org.scalatest.exceptions.TestFailedException: Processors not shutdown for removed listener
15:55:45         at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
15:55:45         at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
15:55:45         at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1389)
15:55:45         at org.scalatest.Assertions.fail(Assertions.scala:1091)
15:55:45         at org.scalatest.Assertions.fail$(Assertions.scala:1087)
15:55:45         at org.scalatest.Assertions$.fail(Assertions.scala:1389)
15:55:45         at kafka.server.DynamicBrokerReconfigurationTest.verifyRemoveListener(DynamicBrokerReconfigurationTest.scala:1178)
15:55:45         at kafka.server.DynamicBrokerReconfigurationTest.testAddRemoveSaslListeners(DynamicBrokerReconfigurationTest.scala:1057)
```

apovzner committed:
clean up connection rate sensors

apovzner commented:
Hi @rajinisivaram, the test failure turned out to be a bug where I did not remove connection rate sensors on listener removal. I fixed the code and tests now pass.

rajinisivaram commented:
ok to test

rajinisivaram commented:
retest this please

rajinisivaram commented:
ok to test

rajinisivaram commented:
@apovzner Thanks for the updates, LGTM. Merging to trunk.
